src/api.ts

```diff
@@ -21,18 +32,18 @@ export interface BlockIndex extends BlockHeader {
 
 export interface RootsReader {
   version: number
-  getRoots: () => Promise<CID[]>
+  getRoots: () => Promisable<CID[]>
 }
 
 export interface BlockIterator extends AsyncIterable<Block> {}
 
 export interface CIDIterator extends AsyncIterable<CID> {}
 
 export interface BlockReader {
-  has: (key: CID) => Promise<boolean>
-  get: (key: CID) => Promise<Block | undefined>
-  blocks: () => BlockIterator
-  cids: () => CIDIterator
+  has: (key: CID) => Promisable<boolean>
+  get: (key: CID) => Promisable<Block | undefined>
+  blocks: () => AnyIterable<Block>
+  cids: () => AnyIterable<CID>
```

Comment:
Its the interface for all the readers, the impls will make it sync or async. The user will have the specific return type and if the user needs a specific type for inputs they can use the class type meaning they only support sync or async not both.
If they support both they can use this type like the tests.

Reply:
I think it's useful to have both interfaces one that isn't prescribing sync / async and separate one that is explicitly sync like we do in https://github.com/multiformats/js-multiformats/blob/167f135eb00291e465a811c849030c0816cac3ea/src/hashes/interface.ts#L34-L72

Referring to implementations is not good idea, because then I can't bring another compatible sync implementation as a drop-in replacement. Also every time library changes even minor updates will break types because class is different even if the implementation exactly the same. Interfaces don't have that problem.

For the reasons above I'd also recommend having second interface that is `BlockReaderSync` like @alanshaw suggested. That way your class can say it implements it and my function can say it wants `BlockReaderSync` as argument without having to introduce dependency on js-car.